### PR TITLE
add @pure docstring and example

### PR DESCRIPTION
**Pull request by miguelraz**
I aggregated Chris Rackauckas's and mbauman's discussion form here

https://discourse.julialang.org/t/pure-macro/3871/3

Into the docstring.

### CODE REVIEW

File: base/expr.jl [hunk 1]
```diff
@@ -184,6 +184,57 @@ macro noinline(ex)
...
+
+Annotates a function to ease type inference by strictly specifying the
+output is completely determined by the input.
+
+Note: Misuse can lead to regressions.
+
+Do not use if the function involved:
+
+- Involves globals, pointers
+- Recurses
```
Comment (vtjnash) [inline base/expr.jl]: What’s wrong with recursion?

File: base/expr.jl [hunk 2]
```diff
@@ -184,6 +184,57 @@ macro noinline(ex)
     esc(isa(ex, Expr) ? pushmeta!(ex, :noinline) : ex)
 end
 
+"""
+    @pure
+
+Annotates a function to ease type inference by strictly specifying the
+output is completely determined by the input.
+
+Note: Misuse can lead to regressions.
```
Comment (vtjnash) [inline base/expr.jl]: I don’t know what regression means. This should just say “Usage can easily lead to whole program corruption or crashes and should be avoided.”

File: base/expr.jl [hunk 3]
```diff
@@ -184,6 +184,57 @@ macro noinline(ex)
...
+
+Do not use if the function involved:
+
+- Involves globals, pointers
+- Recurses
+- Does not return exactly (`===`) the same result for the same input
+- Gets its methods extended after it is called
+
```
Comment (vtjnash) [inline base/expr.jl]: - Uses dispatch on one of its arguments.

File: base/expr.jl (hunk 2 above)
Comment (StefanKarpinski) [reply base/expr.jl]: Did you mean to say "Usage" rather than "Misuse"? 😂

File: base/expr.jl (hunk 3 above)
Comment (miguelraz) [reply base/expr.jl]: Thanks!
Clarifying: one, or any of its arguments?

File: base/expr.jl [hunk 4]
```diff
@@ -184,6 +184,56 @@ macro noinline(ex)
...
+Body:
+  begin 
+      return ((Core.apply_type)(Main.Discrete, false, false)::Type{Discrete{_,_}} where _ where _)()::Discrete{_,_} where _ where _
+  end::Discrete{_,_} where _ where _
+
+julia> struct Discrete2{apply_map,scale_by_time} end
+
+julia> Base.@pure Discrete2(;apply_map=false,scale_by_time=false) = Discrete{apply_map,scale_by_time}()
+Discrete
```
Comment (vtjnash) [inline base/expr.jl]: delete me (duplicate code)

File: base/expr.jl [hunk 5]
```diff
@@ -184,6 +184,56 @@ macro noinline(ex)
...
+
+Do not use if the function involved:
+
+- Involves globals, pointers
+- Recurses
+- Does not return exactly (`===`) the same result for the same input
+- Gets its methods extended after it is called
+- Uses dispatch on one of its arguments
+
+### Example
```
Comment (vtjnash) [inline base/expr.jl]: This example isn't wrong, but it will almost certainly be made obsolete by #24362

File: base/expr.jl (hunk 3 above)
Comment (vtjnash) [reply base/expr.jl]: Yes, any.

File: base/expr.jl [hunk 6]
```diff
@@ -184,6 +184,56 @@ macro noinline(ex)
     esc(isa(ex, Expr) ? pushmeta!(ex, :noinline) : ex)
 end
 
+"""
+    @pure
+
+Usage can easily lead to whole program corruption or crashes and should be avoided
+by beginners.
```
Comment (vtjnash) [inline base/expr.jl]: "by all users."

File: base/expr.jl (hunk 1 above)
Comment (miguelraz) [reply base/expr.jl]: It came up during a Discourse post and @mbauman mentioned the restriction.
https://discourse.julialang.org/t/pure-macro/3871

**Comment by mschauer** [comment]
The definition of what `@pure` actually does got lost during a later commit. (Even if use is discouraged, the doc-string should inform about use and purpose.)

**Comment by vtjnash** [comment]
Usage is “do not use”: what’s not to like 😜?

File: base/expr.jl (hunk 1 above)
Comment (vtjnash) [reply base/expr.jl]: typejoin is pure and does recursion - seems to be doing OK

**Comment by yurivish** [comment]
I think the name of the macro is what makes it so tempting to potential users, even if documented as “really, don’t use this.”. Pure is such a nice word, and people think they know what it means. It’s would be a lot of churn, but is renaming the function to something less exciting something within the realm of possibility for 1.0?

**Comment by vtjnash** [comment]
> within the realm of possibility for 1.0?

Sure. I reserve the right to change the name whenever anyone feels so inclined.

**Comment by StefanKarpinski** [comment]
I've proposed `@hyperpure` previously.

File: base/expr.jl (hunk 1 above)
Comment (miguelraz) [reply base/expr.jl]: Thanks! Will ammend.

**Comment by miguelraz** [comment]
I can put in the work to get `@hyperpure` pre 1.0 (this means I can work on this during feature freeze, right?), but would very much like at least a canonical example of where `@hyperpure` is actually useful.
Any help/ pointers?

**Comment by StefanKarpinski** [comment]
Just so you're aware, `@pure` isn't exported so this can be changed at any time.

**Comment by miguelraz** [comment]
That would explain why I couldn't find examples of it in Base -.-

**Comment by YingboMa** [comment]
```julia
-> % rg '@pure'
NEWS.md
374:    accessible via the `@pure` constructor `Val(c)`. Functions are defined as

test/inference.jl
400:Base.@pure function fpure(a=rand(); b=rand())
416:# Make sure @pure works for functions using the new syntax
417:Base.@pure (fpure2(x::T) where T) = T
709:Base.@pure b20704(@nospecialize(x)) = f20704(x)
720:Base.@pure g20704(::Int) = 1
725:Base.@pure c20704() = (f20704(1.0); 1)
729:Base.@pure function a20704(x)
980:Base.@pure plus1(x) = x + 1

base/inference.jl
78:    actual::Bool  # if true, we obtained `val` by actually calling a @pure function
1599:@pure function type_typeof(@nospecialize(v))
3606:        # TODO: Improve this analysis; if a function is marked @pure we should really

base/broadcast.jl
153:Base.@pure function BroadcastStyle(a::A, b::B) where {A<:AbstractArrayStyle{M},B<:AbstractArrayStyle{N}} where {M,N}

base/irrationals.jl
31:@pure function convert(::Type{Rational{T}}, x::AbstractIrrational) where T<:Integer
47:@pure function (t::Type{T})(x::AbstractIrrational, r::RoundingMode) where T<:Union{Float32,Float64}
80:@pure function rationalize(::Type{T}, x::AbstractIrrational; tol::Real=0) where T
83:@pure function lessrational(rx::Rational{<:Integer}, x::AbstractIrrational)
84:    # an @pure version of `<` for determining if the rationalization of

base/namedtuple.jl
135:@pure function sym_in(x::Symbol, itr::Tuple{Vararg{Symbol}})
142:@pure function merge_names(an::Tuple{Vararg{Symbol}}, bn::Tuple{Vararg{Symbol}})
152:@pure function merge_types(names::Tuple{Vararg{Symbol}}, a::Type{<:NamedTuple}, b::Type{<:NamedTuple})

base/linalg/linalg.jl
20:    @propagate_inbounds, @pure, reduce, typed_vcat

base/linalg/rowvector.jl
67:@pure check_types(::Type{T1}, ::Type{T2}) where {T1,T2} = T1 === transpose_type(T2) ? nothing :

doc/src/manual/types.md
1347:julia> Base.@pure Val(x) = Val{x}()
```

There are several places in Julia where `@pure` is used, so you need to change them too.

**Comment by fredrikekre** [comment]
Is it really necessary to change the name? Why would that stop people from using it (incorrectly)?

**Comment by StefanKarpinski** [comment]
> Is it really necessary to change the name? Why would that stop people from using it (incorrectly)?

The problem is that a "pure" function has a more familiar sense which this is stricter than. People refer to functions that only depend on their arguments and not on any other mutable program state as being pure. Now technically, if you include the method tables of functions as mutable program state (which it is), then the `@pure` macro does have this meaning, however, people aren't accustomed to considering method tables of all the function they use as part of the definition of `@pure`. Moreover, it's almost impossible for the behavior of a non-trivial method definition in Julia not to depend in some way on some method tables. So the idea behind calling it `@hyperpure` is to serve as a warning to people that it probably doesn't mean what they think it means. Then again, it's technically accurate and it's not an exported macro, so maybe just documenting the method table part is sufficient.

**Comment by mbauman** [comment]
I would still strongly support this change.

```
$ ack --type=julia '@pure' ALL_PACKAGES | wc -l
396
```

Just in my skimming through, I'd imagine that's at least 300 too many.

**Pull request closed without merging**